Generic/Syntax: add support for inspecting code passed via STDIN

rodrigoprimo opened the pull request:
# Description

This PR improves the `Generic.PHP.Syntax` sniff to make it work when code is passed via STDIN. Before, any code passed this way would cause a false negative as the sniff was not taking into account that `$phpcsFile->getFilename()` might return `STDIN` instead of an actual file name.

Refer to #915 for additional context and testing instructions.

## Suggested changelog entry

Fixed: the `Generic.PHP.Syntax` sniff now handles code passed via `STDIN` correctly

## Related issues/external references

Fixes #915


## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix _(non-breaking change which fixes an issue)_
- [ ] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [ ] Documentation improvement


## PR checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [x] I have added tests to cover my changes.
- [x] I have verified that the code complies with the projects coding standards.
- [ ] \[Required for new sniffs\] I have added XML documentation for the sniff.

<!--
============================================================================================
Please make sure your pull request passes all continuous integration checks!

PRs which are failing their CI checks will likely be ignored by the maintainers.

Small PRs using atomic, descriptive commits are hugely appreciated as it will make
reviewing your changes easier for the maintainers.
============================================================================================
-->

jrfnl commented:
@rodrigoprimo I believe there is still something wrong in the command as the tests do not pass on Windows.

Aside from that, could you please rebase this branch on the current `master` ?

rodrigoprimo commented:
> Aside from that, could you please rebase this branch on the current master ?

I just rebased without changes, and now `remark` is passing.

> @rodrigoprimo I believe there is still something wrong in the command as the tests do not pass on Windows.

I will take a look to understand why the tests are failing on Windows. When I was creating this PR, I checked that the `Quicktest` workflow was passing, but I completely forgot that the full test suite is not executed on Windows in this workflow.

rodrigoprimo commented:
@jrfnl, I added the `@group Windows` to the new test and fixed the issue on Windows.

As far as I could research, it is not straightforward to reliably pipe commands and escape characters when echoing on Windows, so I opted to create a temporary file instead when on this OS, and the file content is passed via `STDIN`. What do you think of this approach?

rodrigoprimo commented:
Good point, I failed to consider that PHPCS currently doesn't support `STDIN` on Windows. I went ahead, kept the original commit, dropped the two that I had added to make the sniff work when handling `STDIN` on Windows, and added a new one to skip the `STDIN` test on this OS.

jrfnl commented on src/Standards/Generic/Sniffs/PHP/SyntaxSniff.php:
```diff
@@ -72,4 +71,25 @@ public function process(File $phpcsFile, $stackPtr)
 
 
+    /**
+     * Returns the command used to lint PHP code. Uses a different command when the content is
+     * provided via STDIN.
```
```suggestion
     * Returns the command used to lint PHP code.
     *
     * Uses a different command when the content is provided via STDIN.
```

jrfnl commented on src/Standards/Generic/Tests/PHP/SyntaxUnitTest.php:
```diff
@@ -60,4 +63,107 @@ public function getWarningList()
 
 
+    /**
+     * Test the sniff checks syntax when file contents are passed via STDIN. Doesn't run on Windows
+     * as PHPCS currently doesn't support STDIN on this OS.
```
```suggestion
     * Test the sniff checks syntax when file contents are passed via STDIN.
     *
     * Note: this test doesn't run on Windows as PHPCS currently doesn't support STDIN on this OS.
```

jrfnl reviewed (approved):
Thanks for these updates @rodrigoprimo ! LGTM. Will merge once the build has passed.
